### ISSUE TITLE
First "F3" not working.

### ISSUE THREAD

**Issue by bleehome** [mabe02/lanterna#430]
We use key "KeyType.F3" for shortcut of searching action, and the first press of F3 has no effect.
![image](https://user-images.githubusercontent.com/1271744/65575905-74938900-dfa3-11e9-8407-9d2432d898f5.png)
![image](https://user-images.githubusercontent.com/1271744/65575938-88d78600-dfa3-11e9-9ca2-7af4b5cfe28e.png)

F3 is used for cursor position reporting?

**Comment by bleehome** [comment]
I set a position to lastReportedCursorPosition by :

![image](https://user-images.githubusercontent.com/1271744/65577764-0ea90080-dfa7-11e9-82c8-14668daadd85.png)

but it is ugly.

**Comment by avl42** [comment]
You're right, that it is unfortunate, that designers of terminals came up with the questionable "idea" to send the same sequence for cursor-position queries as for user typing a specific key.

Lanterna tries hard not to accidentally misinterpret a cursor-report for a spurious F3, but there is some room for error the other way, if the F3 key is already hit, before the application shows up on the screen.

**Comment by avl42** [comment]
Something could be changed to at least get un"shift"ed,un"control"ed,un"alt"ed F3, which (at least on some terminals I checked) is distinguishable from a cursor-report.   It would mean adding a flag to KeyStroke, and set it for F3, depending on whether it came from "^[OR" or from "^[[1;1R".   Maybe, even "^[[1;1R" is a special case that doesn't ever come from terminals, thus could be special-cased to not give an F3, and then we could restrict the reverse-conversion (second screenshot).

@mabe02 ?

**Comment by mabe02** [comment]
Yikes, I had even forgotten about this... So @bleehome is this consistently never working the first time? Would it be possible for you to put together a small sample test to reproduce this?

**Comment by avl42** [comment]
The problem here is that his F3 keypress comes before the requested
response for terminal cursor position - for determining terminal size, and
because F3-escape codes may look like terminal position responses, lanterna
tends to map apparent F3 keys back to position reports, if one is expected.

mabe02 <notifications@github.com> schrieb am Sa., 16. Nov. 2019 08:57:

**Comment by bleehome** [comment]
> Yikes, I had even forgotten about this... So @bleehome is this consistently never working the first time? Would it be possible for you to put together a small sample test to reproduce this?
I made a very simple  demo but it hangs on waiting for ReportedCursorPosition. so f3 works every first time.
 maybe my own "KeyDecodingProfile" cause this problem, it used to prevent hangs on ssh start start

**Comment by mabe02** [comment]
I've written a special test to try to reproduce this but have been unsuccessful so far. Can you please check the latest `release/3.0` branch and run the `com.googlecode.lanterna.gui2.InputUITest` and see if the issue happens there for you?

**Comment by avl42** [comment]
I can reproduce it.

First of all, it is specific to text-Terminals, thus: not swing (because swing doesn't deal with ANSI escape codes)

So, you need to change the Debug-configuration for com.googlecode.lanterna.gui2.InputUITest to pass options "--telnet-port" and "--no-swing" and connect to it with telnet from a terminal.  (or run the test directly from command line, instead of from eclipse)

When you run it, (or when you connect to localhost 1024 from a terminal) **quickly** press "F3" Key, and then you get this:

![Bildschirmfoto vom 2019-11-17 10-35-46](https://user-images.githubusercontent.com/5255346/69005765-e5fa1300-0926-11ea-971e-85b63863bd47.png)

You see, that the F3-key got mistaken for the cursorlocation-report, and then the real cursorlocation appears as keypress.

I'll try to find a fix for it.

**Comment by avl42** [comment]
I've got a quick-hack that solves it for bare F3, but breaks it, if the terminal ever is resized to 1x1 size.

Maybe @bleehome may patch his local lanterna source, and add line `if (col==1) { return null; }` into method `ScreenInfoCharacterPattern.trytoAdopt` line 55 right after initialization of `col`. In the meantime I will look further, for a slightly better fix, that will work also for 1x1 sized terminals.  But there is **no way** to fix it for Shift'ed, Ctrl'ed, or Alt'ed F3, unfortunately. I hope bare F3 is all you need.
